cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeMonitor.java

```diff
@@ -144,19 +176,25 @@ private int processResponse(Map<String, List<MonitorMessage>> monitorResponses)
       if (monitorResponse.getKey().equals(Constants.AppFabric.PROGRAM_STATUS_EVENT_TOPIC)) {
         setIsRuntimeInactive(monitorResponse.getValue());
       }
-      publish(monitorResponse.getKey(), monitorResponse.getValue());
+      String topic = getTopic(monitorResponse.getKey());
+      publish(monitorResponse.getKey(), topic, monitorResponse.getValue());
       count += monitorResponse.getValue().size();
     }
     return count;
   }
 
-  private void publish(String topicConfig, List<MonitorMessage> messages) throws Exception {
+  private String getTopic(String topicConfig) {
```

Comment:
Shouldn't need to do this every time. You can figure out the topic on the master side when you are constructing the `topicsToRequest` map. You can store it there.